packages/molecule/src/codec.ts

```diff
@@ -0,0 +1,233 @@
+        mol.Codec<any, any>
+      > = unionCodecs.reduce(
+        (codecRecord, [fieldName, _fieldId, fieldCodec]) =>
+          Object.assign(codecRecord, { [fieldName]: fieldCodec }),
+        {},
+      );
+      const unionFieldIds: Record<string, number> = unionCodecs.reduce(
+        (idMap, [fieldName, fieldId, _fieldCodec]) =>
+          Object.assign(idMap, { [fieldName]: fieldId }),
+        {},
+      );
+
+      codec = mol.union(unionFieldsCodecs, unionFieldIds);
+      break;
+    }
+    case "table": {
+      const tableFields = molTypeDefinition.fields;
+      const tableCodecs: Record<string, mol.Codec<any, any>> = {};
+      tableFields.forEach((field) => {
+        if (field.type === BYTE) {
+          tableCodecs[field.name] = mol.Byte;
+        } else {
+          const itemCodec = outputCodecRecord[field.type];
+          if (!itemCodec) {
+            throw new Error(
+              `Codec not found for item type: ${field.type} in table type: ${molTypeDefinition.name}`,
+            );
+          }
+          tableCodecs[field.name] = itemCodec;
+        }
+      });
+      codec = mol.table(tableCodecs);
+      break;
+    }
+    case "struct": {
+      const structFields = molTypeDefinition.fields;
+      const structCodecs: Record<string, mol.Codec<any, any>> = {};
+      structFields.forEach((field) => {
+        if (field.type === BYTE) {
+          structCodecs[field.name] = mol.Byte;
+        } else {
+          const itemCodec = outputCodecRecord[field.type];
+          if (!itemCodec) {
+            throw new Error(
+              `Codec not found for field type: ${field.type} in struct type: ${molTypeDefinition.name}`,
+            );
+          }
+          structCodecs[field.name] = itemCodec;
+        }
+      });
+      codec = mol.struct(structCodecs);
+      break;
+    }
+  }
+  if (!codec) {
+    throw new Error(`Codec not found for type: ${molTypeDefinition.name}`);
+  }
+  return codec;
+}
+
+/**
+ * Creates a CodecRecord from an array of molecule type definitions.
+ *
+ * Takes an array of molecule type definitions and converts each one into a codec
+ * that can be used for serialization/deserialization, and create a CodecRecord indexed by the type name.
+ *
+ * It would handle the dependencies of the type definitions, and the dependencies of the dependencies.
+ *
+ * @param molTypeDefinitions - The molecule type definitions to convert. Should be obtained by grammar.parse(molSchemaSrc).declares.
+ * @param extraReferences - Optional pre-defined codecs that can be referenced instead of creating new ones
+ * @returns CodecRecord indexed by the type name
+ */
+export function toCodecRecord(
+  molTypeDefinitions: MolTypeDefinition[],
+  extraReferences?: CodecRecord,
+): CodecRecord {
+  const outputCodecRecord: CodecRecord = extraReferences || {};
+  const processed = new Set<string>();
+
+  function process(definition: MolTypeDefinition) {
+    if (processed.has(definition.name)) return;
```

Comment:
`{ ... }` is always recommended for `if` - too many historical catastrophes have been caused by this.